### PR TITLE
[DDP] Add API to get model parameters in hook

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61637**

To support running optimizer as a communication hook, add API to
retrieve the model parameters.

The API returns a `List[tensor]` that contains model parameters for the bucket. It can be used as follows:
```
per_param_grad_tensors = bucket.get_per_parameter_tensors()
model_params_for_bucket = bucket.get_model_params_for_bucket()
# Now gradient per_param_grad_tensors[i] is the gradient tensor for model_params_for_bucket[i].
```

This provides a way for comm. hook developer to retrieve model parameters within a hook. In the next diffs, we will use this to run optimizer as a DDP comm. hook.

Differential Revision: [D29691418](https://our.internmc.facebook.com/intern/diff/D29691418/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -301,6 +301,11 @@ to apply layer-wise operations.
 Returns:
     A list of ``torch.Tensor``. Each tensor in the list corresponds to a parameter.
 )")
+      .def(
+          "get_grad_index_to_variable_mapping",
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:305]: curious, is the param index sth that users must know? Will it work if we just expose one single API `bucket.parameters()` that returns references to model parameters with the `.grad` field properly populated?

**Review by mrshenli — COMMENTED** [review]
cc @andwgu

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -7513,3 +7513,39 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
...
+        )
+        @skip_if_lt_x_gpu(2)
+        @skip_if_rocm
+        def test_hook_grad_idx_to_bucket_mapping(self):
+            # Test hook to ensure get_grad_index_to_variable_mapping works appropriately.
+            def _test_hook(hook_state, bucket: dist.GradBucket) -> torch.futures.Future:
+                # Run allreduce and wait for it to complete
+                allreduce_fut = default._allreduce_fut(None, bucket.get_tensor())
+                allreduce_fut.wait()
+                per_param_grad_tensors = bucket.get_per_parameter_tensors()
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:7529]: Would I be correct if I assume these tensors are views on the bucket?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 3]
```diff
@@ -7513,3 +7513,39 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
...
+                idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
+                # Verify mapping is correct by ensuring param.grad is equal to
+                # the grad we get using the mapping.
+                for grad_tensor_idx, model_param in idx_to_model_params.items():
+                    self.assertEqual(model_param.grad, per_param_grad_tensors[grad_tensor_idx])
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/distributed_test.py:7534]: If the above assumption is correct, how does this work? Since the test doesn't enable `grad_as_bucket_view`, I assume `param.grad` are not bucket views, and the allreduce result are not yet copied back to `param.grad`. I might miss sth, why do we expect `param.grad` and bucket views to be the same value?

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/c10d/init.cpp:305]: Above might only work when `grad_as_bucket_view` is enabled. Curious, when `grad_as_bucket_view=False`, how will optimizer work in the communication hook? Are we expecting users to manually populate `param.grad` before calling `optimizer.step()`? If so, will it work if we expose `bucket.parameters` and `bucket.grads` as two separate APIs, where parameters() return references to model parameters and `grads()` return bucket views?

File: torch/csrc/distributed/c10d/comm.hpp [hunk 4]
```diff
@@ -51,6 +53,15 @@ class TORCH_API GradBucket {
   // parameter.
   std::vector<at::Tensor> getPerParameterTensors() const;
 
+  // Returns mapping of gradient index in bucket to model parameter. Practically
+  // this can be used in conjunction with getPeraParameterTensors() as follows:
```
Comment (andwgu) [inline torch/csrc/distributed/c10d/comm.hpp:57]: nit: `getPeraParameterTensors()` -> `getPerParameterTensors()`

File: torch/csrc/distributed/c10d/comm.hpp [hunk 5]
```diff
@@ -51,6 +53,15 @@ class TORCH_API GradBucket {
...
+  // auto grads = getPerParameterTensors(); // grad tensors for this bucket
+  // auto mapping = getGradIndexToVariable();
```
Comment (andwgu) [inline torch/csrc/distributed/c10d/comm.hpp:59]: nit: I think perhaps `getGradIndexToVariable()` was your old version of the function name, and this should be `auto mapping = getGradIndexToVariableMapping()` now.

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/init.cpp:305]: Regarding bucket.parameters and then using param.grad field, that might be hard since we will be running optimizer as a `then()` callback to gradient reduction, but `param.grad` is not populated until `Reducer::finalize_backward` is called. So if we access `param.grad` it will be a stale gradient in the comm. hook.

> Curious, when grad_as_bucket_view=False, how will optimizer work in the communication hook? Are we expecting users to manually populate param.grad before calling optimizer.step()? 

In both cases, we are planning to do something like the following:
```
grad_tensors = bucket.get_per_parameter_tensors()
idx_to_param = bucket.get_idx_to_params()
for i in range(len(grad_tensors)):
    grad, model_param = grad_tensors[i], idx_to_param[i]
    functional_optimizer.step(model_param, grad)
```

For this to work we'll need functional optimizers that operate on explicit gradients which we've created a version of for the RPC framework.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py:7529]: Yep, that is correct. There's no copying/cloning of tensors going on here.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 3 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py]: This is very strange because we're also not calling `zero_grad` anywhere so theoretically the grad field should accumulate. Looking into it...

Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py]: @mrshenli This appears to pass because the gradients are so close to 0 (on the order of 1e-8) that when I print them out they are indeed different, but `self.assertEqual` default tolerance is too generous marking them as equal.

If we scale up the gradients this test indeed fails. This makes sense because during backwards the grads have not yet been copied back to param.grad field. Although, the main use case here is to just verify we can get the appropriate gradient for the parameter, so we can probably just do shape check when `grad_as_bucket_view=False` and grad check when `grad_as_bucket_view=True`?

File: torch/csrc/distributed/c10d/comm.hpp [hunk 6]
```diff
@@ -21,12 +21,14 @@ class TORCH_API GradBucket {
...
-      const std::vector<c10::IntArrayRef>& sizes_vec)
+      const std::vector<c10::IntArrayRef>& sizes_vec,
+      const std::unordered_map<size_t, at::Tensor>& bucket_idx_to_variable)
       : index_(index),
         tensor_(tensor),
         offsets_(offsets),
         lengths_(lengths),
-        sizes_vec_(sizes_vec) {}
+        sizes_vec_(sizes_vec),
+        bucket_idx_to_variable_(bucket_idx_to_variable) {}
```
Comment (zhaojuanmao) [inline torch/csrc/distributed/c10d/comm.hpp:31]: could we just call getVariables() to return as list of variables? as the idx is the same as idx in getPerParameterTensor() anyway?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 7]
```diff
@@ -7513,3 +7513,48 @@ def test_ddp_bwd_with_retain_graph_static_single_fwd(self):
...
+                allreduce_fut.wait()
+                per_param_grad_tensors = bucket.get_per_parameter_tensors()
+                idx_to_model_params = bucket.get_grad_index_to_variable_mapping()
+                # Verify mapping is correct by ensuring param.grad is equal to
+                # the grad we get using the mapping.
+                m = {i : p.shape for i, p in idx_to_model_params.items()}
+                for i in range(len(per_param_grad_tensors)):
+                    grad = per_param_grad_tensors[i]
+                    model_param = idx_to_model_params[i]
+                    self.assertEqual(model_param.grad, grad)
```
Comment (zhaojuanmao) [inline torch/testing/_internal/distributed/distributed_test.py:7537]: I think the question is: grad is averaged grad, model_param.grad is old grad, why are they equal here?

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/c10d/init.cpp:305]: Will sth like this work:

```python
for param, grad in zip(bucket.parameters(), bucket.gradients()):
    # param is the model parameter, whose .grad field is not populated yet
    # grad is a view to the bucket offset, and it already contains the allreduced grad. 
    functional_optimizer.step(model_param, grad)
```

Comment (mrshenli) [reply torch/csrc/distributed/c10d/init.cpp:305]: Above code means that `parameters()` and `gradients()` must match in order, but it should be fine as they both belong to the same bucket. And it will also avoid engaging with param index.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 7 above)
Comment (rohan-varma) [reply torch/testing/_internal/distributed/distributed_test.py]: It was an issue because the numerics were too small and tolerance was high enough to make them equal. Should be resolved now that we check for shape when grad_as_bucket_view=False.

**Review by mrshenli — COMMENTED** [review]
(This is prior to this PR) APIs on GradientBucket is a bit verbose. Looks like they can be simplified to, 

`get_index` -> `index` 

`get_tensor` -> `buffer` 

`set_tensor` -> `set_buffer` 

`is_the_last_bucket_to_allreduce` -> `is_last`,

`get_per_parameter_tensors` -> `parameters`, 

new API -> `gradients`. 

Since CommHook is still beta and will soon become stable, shall we take this chance to do one round of discussion and revamp to those APIs for simplicity?


--------------------

Please feel free to apply the API revamp to followup PRs

File: torch/testing/_internal/distributed/distributed_test.py (hunk 3 above)
Comment (mrshenli) [reply torch/testing/_internal/distributed/distributed_test.py]: Yep, that looks good to me.

**Comment by rohan-varma** [comment]
> (This is prior to this PR) APIs on GradientBucket is a bit verbose.
> Please feel free to apply the API revamp to followup PRs

Sure, I'll send a follow up diff with those changes.

One note:

> `get_per_parameter_tensors` -> `parameters`,
> 
> new API -> `gradients`.

I think it should basically be the other way around, as `get_per_param_tensors` actually returns grads and this API returns the model params. Name is sort of confusing.

File: torch/_C/_distributed_c10d.pyi [hunk 8]
```diff
@@ -25,6 +25,7 @@ class GradBucket:
     def get_per_parameter_tensors(self) -> List[Tensor]: ...
     def is_the_last_bucket_to_allreduce(self) -> bool: ...
     def set_tensor(self, tensor: Tensor) -> None: ...
+    def get_model_params_for_bucket(self) -> List[Tensor]: ...
```
Comment (SciPioneer) [inline torch/_C/_distributed_c10d.pyi:28]: NI: I think we can just call this `get_params`, since all the data in this `GradBucket` class is for the bucket only. No need to add extra `for_bucket`.

Comment (rohan-varma) [reply torch/_C/_distributed_c10d.pyi:28]: I'll call it `get_model_params` for now, otherwise could be confusing with existing `get_per_parameter_tensors`, which is actually gradients, which will be renamed in a follow up diff

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 9]
```diff
@@ -845,17 +848,53 @@ void Reducer::all_reduce_bucket(Bucket& bucket) {
...
+  std::vector<at::Tensor> variables_for_bucket;
+  variables_for_bucket.reserve(bucket.variable_indices.size());
+  for (const auto& variable_index : bucket.variable_indices) {
+    auto& replica = bucket.replicas[0];
+    // Grab bucket index where gradient is located using variable_locators_.
+    auto& bucket_index_for_variable = variable_locators_[variable_index];
+    // Grab the actual model parameter.
+    auto& variable =
+        replica.variables[bucket_index_for_variable.intra_bucket_index];
+    variables_for_bucket.emplace_back(variable);
```
Comment (SciPioneer) [inline torch/csrc/distributed/c10d/reducer.cpp:883]: Will you still access `replica.variables`?

**Review by wayi1 — APPROVED** [review]

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp]: Probably this can be a follow-up refactoring PR.

**Comment by SciPioneer** [comment]
> > (This is prior to this PR) APIs on GradientBucket is a bit verbose.
> > Please feel free to apply the API revamp to followup PRs
> 
> Sure, I'll send a follow up diff with those changes.
> 
> One note:
> 
> > `get_per_parameter_tensors` -> `parameters`,
> > new API -> `gradients`.
> 
> I think it should basically be the other way around, as `get_per_param_tensors` actually returns grads and this API returns the model params. Name is sort of confusing.

Correct, previously we use "tensor" to refer to the gradient data. Now we add parameters, so better to give them better names.
